# Merging all data from different sources and writing csv files for descriptive and modelling parts

Install required packages for geopa

In [1]:
# import needed libraries
import pandas as pd
import numpy as np

In [2]:
# read all data obtained from the weather_data and MRT_calculation notebooks
weather_final_01 = pd.read_csv('FINAL_WEATHER_MERGED_01', index_col=0)
weather_final_05 = pd.read_csv('FINAL_WEATHER_MERGED_05', index_col=0)
weather_final_09 = pd.read_csv('FINAL_WEATHER_MERGED_09', index_col=0)
weather_final_13 = pd.read_csv('FINAL_WEATHER_MERGED_13', index_col=0)
weather_final_17 = pd.read_csv('FINAL_WEATHER_MERGED_17', index_col=0)
weather_final_21 = pd.read_csv('FINAL_WEATHER_MERGED_21', index_col=0)

In [3]:
# concatenating all weather dfs to a large df
frames = [weather_final_01, weather_final_05, weather_final_09, weather_final_13, weather_final_17, weather_final_21]

weather_df = pd.concat(frames)
weather_df.shape

(3891096, 33)

In [4]:
# writing concatenated df to csv. This df is used for descriptive statistics
weather_df.to_csv('weather_df.csv')

## Used in Notebook of the descriptive part --> Descriptive.ipnyb

## From here on, the final dataset for the modelling will be set up

In [5]:
# showing unique latitudes
weather_df['latitude'].unique()

array([54.25, 53.5 , 52.  , 52.75, 43.  , 39.25, 55.  , 42.25, 41.5 ,
       40.75, 40.  , 38.5 , 37.75, 58.  , 57.25, 55.75, 37.  , 56.5 ,
       50.5 , 48.25, 51.25, 47.5 , 46.75, 49.  , 46.  , 45.25, 44.5 ,
       43.75, 49.75, 61.  , 61.75, 60.25, 59.5 , 58.75, 62.5 , 63.25,
       64.  , 64.75, 66.25, 65.5 , 67.  , 68.5 , 67.75, 69.25, 70.  ,
       70.75])

In [6]:
# show first instances
weather_df.head()

longitude  latitude  wind_2AM  dew_point_2AM  temperature_2AM  uvb_2AM  \
0      -10.0     54.25     -3.17           7.57            10.66    -0.02   
1      -10.0     54.25     -1.20           8.22             9.81    -0.02   
2      -10.0     54.25     -4.29           6.97             8.52    -0.02   
3      -10.0     54.25     -4.35           6.38             8.56    -0.02   
4      -10.0     54.25     -3.99           7.33             9.03    -0.02   

             lat_long        date  apparent_temperature_2AM  wind_11AM  ...  \
0  ('54.25', '-10.0')  2000-05-01                      8.82       0.16  ...   
1  ('54.25', '-10.0')  2000-05-02                      8.13      -1.53  ...   
2  ('54.25', '-10.0')  2000-05-03                      6.56      -4.15  ...   
3  ('54.25', '-10.0')  2000-05-04                      6.47      -4.60  ...   
4  ('54.25', '-10.0')  2000-05-05                      7.15      -3.51  ...   

   apparent_temperature_11PM  country  NUTS1  NUTS2  NUTS3  MRT_2AM  MRT_11AM  \
0                       8.60       IE    IE0   IE04  IE042     2.96     23.10   
1                       7.14       IE    IE0   IE04  IE042     3.03     23.49   
2                       6.43       IE    IE0   IE04  IE042     3.64     27.86   
3                       7.81       IE    IE0   IE04  IE042     1.90     21.78   
4                       8.86       IE    IE0   IE04  IE042    -0.44     12.57   

   MRT_4PM  MRT_11PM  Week_Number  
0    15.57      3.84           18  
1    25.09      5.60           18  
2    23.63      0.11           18  
3    17.08      0.12           18  
4    14.37      5.40           18  

[5 rows x 33 columns]

# Merge with temperature threshold to find heat wave days

In [7]:
# read the threshold dataframe obtained from the Threshold_Computation.ipnyb
thresholds = pd.read_csv('quantile_thresholds.csv', index_col=0)
thresholds.head(2)

latitude  longitude        d2m        t2m  apparent_temperature
0      37.0      -6.25  20.096707  38.971600             39.232187
1      37.0      -5.50  19.063138  38.051736             38.677543

In [8]:
# rename apparent_temperature to 99th percentile
thresholds = thresholds.rename(columns={"apparent_temperature": "99th_percentile"})

In [9]:
# merge the thresold with the weather data
weather_final = weather_df.merge(thresholds,on=['longitude','latitude'], how='left').drop(columns = ['d2m','t2m'])

In [10]:
# inspect the datframe
weather_final.tail()

longitude  latitude  wind_2AM  dew_point_2AM  temperature_2AM  \
3891091      29.75     61.75      2.28           5.96             6.96   
3891092      29.75     61.75     -0.01           2.83             3.14   
3891093      29.75     61.75     -1.48           3.60             3.85   
3891094      29.75     61.75     -1.20           5.68             6.56   
3891095      29.75     61.75     -1.08           5.78             6.76   

         uvb_2AM            lat_long        date  apparent_temperature_2AM  \
3891091      0.0  ('61.75', '29.75')  2021-09-26                      4.81   
3891092      0.0  ('61.75', '29.75')  2021-09-27                      0.60   
3891093      0.0  ('61.75', '29.75')  2021-09-28                      1.37   
3891094      0.0  ('61.75', '29.75')  2021-09-29                      4.36   
3891095      0.0  ('61.75', '29.75')  2021-09-30                      4.58   

         wind_11AM  ...  country  NUTS1  NUTS2  NUTS3  MRT_2AM  MRT_11AM  \
3891091       1.17  ...       FI    FI1   FI1C  FI1C5     4.13     18.18   
3891092       0.18  ...       FI    FI1   FI1C  FI1C5     0.34     20.55   
3891093      -1.15  ...       FI    FI1   FI1C  FI1C5     3.95     17.04   
3891094      -0.11  ...       FI    FI1   FI1C  FI1C5     4.56     16.11   
3891095      -0.67  ...       FI    FI1   FI1C  FI1C5     4.37     15.13   

         MRT_4PM  MRT_11PM  Week_Number  99th_percentile  
3891091     6.60      0.10           38        26.917881  
3891092     3.04      0.14           39        26.917881  
3891093     6.69      5.08           39        26.917881  
3891094     6.52      5.48           39        26.917881  
3891095     6.30      3.14           39        26.917881  

[5 rows x 34 columns]

In [11]:
# transform date to datetime and get day month and year of date
weather_final['date'] = pd.to_datetime(weather_final['date'])
weather_final['day'] = weather_final.date.dt.day
weather_final['month'] = weather_final.date.dt.month
weather_final['year'] = weather_final.date.dt.year

In [12]:
# inspect info
weather_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3891096 entries, 0 to 3891095
Data columns (total 37 columns):
 #   Column                     Dtype         
---  ------                     -----         
 0   longitude                  float64       
 1   latitude                   float64       
 2   wind_2AM                   float64       
 3   dew_point_2AM              float64       
 4   temperature_2AM            float64       
 5   uvb_2AM                    float64       
 6   lat_long                   object        
 7   date                       datetime64[ns]
 8   apparent_temperature_2AM   float64       
 9   wind_11AM                  float64       
 10  dew_point_11AM             float64       
 11  temperature_11AM           float64       
 12  uvb_11AM                   float64       
 13  apparent_temperature_11AM  float64       
 14  wind_4PM                   float64       
 15  dew_point_4PM              float64       
 16  temperature_4PM            float64  

In [13]:
# new binary column indicating if threshold was exceeded
weather_final['threshold_exceeded'] = np.where(weather_final['apparent_temperature_4PM'] > weather_final['99th_percentile'],1,0)

In [14]:
# inspect data
weather_final.head(5)

longitude  latitude  wind_2AM  dew_point_2AM  temperature_2AM  uvb_2AM  \
0      -10.0     54.25     -3.17           7.57            10.66    -0.02   
1      -10.0     54.25     -1.20           8.22             9.81    -0.02   
2      -10.0     54.25     -4.29           6.97             8.52    -0.02   
3      -10.0     54.25     -4.35           6.38             8.56    -0.02   
4      -10.0     54.25     -3.99           7.33             9.03    -0.02   

             lat_long       date  apparent_temperature_2AM  wind_11AM  ...  \
0  ('54.25', '-10.0') 2000-05-01                      8.82       0.16  ...   
1  ('54.25', '-10.0') 2000-05-02                      8.13      -1.53  ...   
2  ('54.25', '-10.0') 2000-05-03                      6.56      -4.15  ...   
3  ('54.25', '-10.0') 2000-05-04                      6.47      -4.60  ...   
4  ('54.25', '-10.0') 2000-05-05                      7.15      -3.51  ...   

   MRT_2AM  MRT_11AM  MRT_4PM  MRT_11PM  Week_Number  99th_percentile  day  \
0     2.96     23.10    15.57      3.84           18        20.898722    1   
1     3.03     23.49    25.09      5.60           18        20.898722    2   
2     3.64     27.86    23.63      0.11           18        20.898722    3   
3     1.90     21.78    17.08      0.12           18        20.898722    4   
4    -0.44     12.57    14.37      5.40           18        20.898722    5   

   month  year  threshold_exceeded  
0      5  2000                   0  
1      5  2000                   0  
2      5  2000                   0  
3      5  2000                   0  
4      5  2000                   0  

[5 rows x 38 columns]

In [15]:
# show how often the threshold was exceeded
weather_final['threshold_exceeded'].value_counts()

0    3808602
1      82494
Name: threshold_exceeded, dtype: int64

In [16]:
# check if there are duplicates in the data
duplicate = weather_df[weather_final.duplicated(['wind_2AM', 'dew_point_2AM', 'temperature_2AM', 'uvb_2AM', 'apparent_temperature_2AM', 'wind_11AM'])]
  
print("Duplicate Rows based on Name and Age :")
  
# Print the resultant Dataframe
duplicate

<ipython-input-16-38e19741e4ee>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  duplicate = weather_df[weather_final.duplicated(['wind_2AM', 'dew_point_2AM', 'temperature_2AM', 'uvb_2AM', 'apparent_temperature_2AM', 'wind_11AM'])]


Duplicate Rows based on Name and Age :


Empty DataFrame
Columns: [longitude, latitude, wind_2AM, dew_point_2AM, temperature_2AM, uvb_2AM, lat_long, date, apparent_temperature_2AM, wind_11AM, dew_point_11AM, temperature_11AM, uvb_11AM, apparent_temperature_11AM, wind_4PM, dew_point_4PM, temperature_4PM, uvb_4PM, apparent_temperature_4PM, wind_11PM, dew_point_11PM, temperature_11PM, uvb_11PM, apparent_temperature_11PM, country, NUTS1, NUTS2, NUTS3, MRT_2AM, MRT_11AM, MRT_4PM, MRT_11PM, Week_Number]
Index: []

[0 rows x 33 columns]

In [17]:
# set coordinates and date as index
weather_final = weather_final.set_index(['lat_long', 'date'])

In [18]:
# show value counts per NUTS 2
weather_final['NUTS2'].value_counts()

FI1D    255816
SE33    191862
NO07    148104
SE32     84150
SE31     77418
         ...  
UKG1      3366
NL23      3366
UKK4      3366
UKF2      3366
DK02      3366
Name: NUTS2, Length: 249, dtype: int64

#### Some NUTS 2 regions appear to be very big.  For the four largest NUTS 2 regions, NUTS 3 regions will be used to reduce the size of regions.

In [20]:
changed_regions = ['FI1D', 'SE31', 'SE32', 'SE33', 'NO07']

weather_final['NUTS2'] = np.where(weather_final['NUTS2'].isin(changed_regions), weather_final['NUTS3'], weather_final['NUTS2'])
weather_final['NUTS2'].value_counts()

FI1D7    127908
SE332    124542
NO074    100980
NO0A      67320
SE331     67320
          ...  
UKG1       3366
UKK4       3366
UKF2       3366
UKH3       3366
DK02       3366
Name: NUTS2, Length: 260, dtype: int64

In [22]:
# exception for the netherlands as data is only available on the national level for a long time period
exception = ['NL']

weather_final['NUTS2'] = np.where(weather_final['country'].isin(exception), weather_final['country'], weather_final['NUTS2'])
weather_final['NUTS2'].value_counts()


FI1D7    127908
SE332    124542
NO074    100980
NO0A      67320
SE331     67320
          ...  
DEA5       3366
UKJ3       3366
ITI2       3366
BE32       3366
DK02       3366
Name: NUTS2, Length: 253, dtype: int64

#### After the change, Finnish, Norwegian and Swedish NUTS 3 regions still remain big, but the improvement is considered sufficient.

In [23]:
weather_final.head()

longitude  latitude  wind_2AM  dew_point_2AM  \
lat_long           date                                                       
('54.25', '-10.0') 2000-05-01      -10.0     54.25     -3.17           7.57   
                   2000-05-02      -10.0     54.25     -1.20           8.22   
                   2000-05-03      -10.0     54.25     -4.29           6.97   
                   2000-05-04      -10.0     54.25     -4.35           6.38   
                   2000-05-05      -10.0     54.25     -3.99           7.33   

                               temperature_2AM  uvb_2AM  \
lat_long           date                                   
('54.25', '-10.0') 2000-05-01            10.66    -0.02   
                   2000-05-02             9.81    -0.02   
                   2000-05-03             8.52    -0.02   
                   2000-05-04             8.56    -0.02   
                   2000-05-05             9.03    -0.02   

                               apparent_temperature_2AM  wind_11AM  \
lat_long           date                                              
('54.25', '-10.0') 2000-05-01                      8.82       0.16   
                   2000-05-02                      8.13      -1.53   
                   2000-05-03                      6.56      -4.15   
                   2000-05-04                      6.47      -4.60   
                   2000-05-05                      7.15      -3.51   

                               dew_point_11AM  temperature_11AM  ...  MRT_2AM  \
lat_long           date                                          ...            
('54.25', '-10.0') 2000-05-01            8.74             13.70  ...     2.96   
                   2000-05-02            9.74             11.66  ...     3.03   
                   2000-05-03            7.06             10.37  ...     3.64   
                   2000-05-04            7.83             11.56  ...     1.90   
                   2000-05-05            7.76             11.78  ...    -0.44   

                               MRT_11AM  MRT_4PM  MRT_11PM  Week_Number  \
lat_long           date                                                   
('54.25', '-10.0') 2000-05-01     23.10    15.57      3.84           18   
                   2000-05-02     23.49    25.09      5.60           18   
                   2000-05-03     27.86    23.63      0.11           18   
                   2000-05-04     21.78    17.08      0.12           18   
                   2000-05-05     12.57    14.37      5.40           18   

                               99th_percentile  day  month  year  \
lat_long           date                                            
('54.25', '-10.0') 2000-05-01        20.898722    1      5  2000   
                   2000-05-02        20.898722    2      5  2000   
                   2000-05-03        20.898722    3      5  2000   
                   2000-05-04        20.898722    4      5  2000   
                   2000-05-05        20.898722    5      5  2000   

                               threshold_exceeded  
lat_long           date                            
('54.25', '-10.0') 2000-05-01                   0  
                   2000-05-02                   0  
                   2000-05-03                   0  
                   2000-05-04                   0  
                   2000-05-05                   0  

[5 rows x 36 columns]

In [24]:
# define function to count number of consecutive days of threshold exceeded, 
# source: https://stackoverflow.com/questions/43616174/pandas-dataframe-how-to-count-the-number-of-1-rows-in-a-binary-column
def cumsum_bincount(a):  
    # Append 0 & look for a [0,1] pattern. Form a binned array based off 1s groups
    ids = a*(np.diff(np.r_[0,a])==1).cumsum()

    # Get the bincount, index into the count with ids and  mask out 0s
    return a*np.bincount(ids)[ids]

In [25]:
# apply function
weather_final['consecutive_HW_days'] = cumsum_bincount(weather_final.threshold_exceeded.values)

In [26]:
# sort by longest heat wave recorded, it lasted 22 days
pd.set_option('display.max_columns', None)
weather_final.sort_values(by=['consecutive_HW_days'], ascending=False).head()

longitude  latitude  wind_2AM  dew_point_2AM  \
lat_long           date                                                       
('60.25', '20.75') 2014-07-25      20.75     60.25      0.54          19.43   
                   2014-08-03      20.75     60.25     -5.97          17.31   
                   2014-07-24      20.75     60.25      1.52          17.63   
                   2014-08-10      20.75     60.25      4.66          18.36   
                   2014-08-09      20.75     60.25      3.75          17.60   

                               temperature_2AM  uvb_2AM  \
lat_long           date                                   
('60.25', '20.75') 2014-07-25            21.59    -0.02   
                   2014-08-03            21.21    -0.02   
                   2014-07-24            20.97    -0.02   
                   2014-08-10            20.49    -0.02   
                   2014-08-09            20.05    -0.02   

                               apparent_temperature_2AM  wind_11AM  \
lat_long           date                                              
('60.25', '20.75') 2014-07-25                     24.58       0.76   
                   2014-08-03                     23.01      -7.78   
                   2014-07-24                     22.94       1.10   
                   2014-08-10                     22.87       2.32   
                   2014-08-09                     22.02       0.99   

                               dew_point_11AM  temperature_11AM   uvb_11AM  \
lat_long           date                                                      
('60.25', '20.75') 2014-07-25           19.86             21.54  298751.06   
                   2014-08-03           19.09             21.53  124010.25   
                   2014-07-24           17.23             21.33  299977.22   
                   2014-08-10           16.16             20.19  240710.72   
                   2014-08-09           18.60             20.20  278398.20   

                               apparent_temperature_11AM  wind_4PM  \
lat_long           date                                              
('60.25', '20.75') 2014-07-25                      24.80     -0.43   
                   2014-08-03                      24.33     -7.96   
                   2014-07-24                      23.09      0.72   
                   2014-08-10                      21.41      1.56   
                   2014-08-09                      22.72      2.23   

                               dew_point_4PM  temperature_4PM    uvb_4PM  \
lat_long           date                                                    
('60.25', '20.75') 2014-07-25          19.19            21.82  142510.39   
                   2014-08-03          19.76            23.15  129143.94   
                   2014-07-24          19.55            21.00  136514.34   
                   2014-08-10          17.04            19.99   85197.67   
                   2014-08-09          18.07            20.47  119759.12   

                               apparent_temperature_4PM  wind_11PM  \
lat_long           date                                              
('60.25', '20.75') 2014-07-25                     24.67       0.06   
                   2014-08-03                     26.33      -6.19   
                   2014-07-24                     24.07      -0.16   
                   2014-08-10                     21.66      -0.65   
                   2014-08-09                     22.69       2.99   

                               dew_point_11PM  temperature_11PM  uvb_11PM  \
lat_long           date                                                     
('60.25', '20.75') 2014-07-25           18.55             22.29     -0.02   
                   2014-08-03           20.21             23.50     -0.02   
                   2014-07-24           19.29             21.68     -0.02   
                   2014-08-10           15.90             19.31     -0.02   
                   2014-08-09           18.65   

In [27]:
# show value counts, each number in the right column should be divided by the value in the left column to obtain the unque count
# of heat wave lengths, e.g. 22/22 = 1 --> 1 heat wave with length 22 days
weather_final['consecutive_HW_days'].value_counts()

0     3808602
1       25863
2       20830
3       12780
4        8784
5        5805
6        3318
7        1582
11       1023
8         880
9         540
10        390
12        360
13         91
16         64
20         60
17         34
22         22
21         21
18         18
15         15
14         14
Name: consecutive_HW_days, dtype: int64

In [28]:
# indicate whether day is a heat wave day
weather_final['heat_wave_day'] = np.where((weather_final['threshold_exceeded'] == 1) & (weather_final['consecutive_HW_days'] >= 4),1,0)

In [29]:
# show frequency of heat wave days
weather_final['heat_wave_day'].value_counts()

0    3868075
1      23021
Name: heat_wave_day, dtype: int64

In [30]:
# group number of heat wave days by year
weather_final.groupby('year')['heat_wave_day'].value_counts()

year  heat_wave_day
2000  0                176554
      1                   314
2001  0                176797
      1                    71
2002  0                176550
      1                   318
2003  0                173429
      1                  3439
2004  0                176710
      1                   158
2005  0                176598
      1                   270
2006  0                176405
      1                   463
2007  0                175732
      1                  1136
2008  0                176511
      1                   357
2009  0                176864
      1                     4
2010  0                175391
      1                  1477
2011  0                176487
      1                   381
2012  0                176030
      1                   838
2013  0                176182
      1                   686
2014  0                175345
      1                  1523
2015  0                174848
      1                  2020
2016  0             

## Recorded heat waves per region and year

In [31]:
# write concatenated data in csv
#weather_final.to_csv('Heat_wave.csv')

In [32]:
# show all NUTS regions
weather_final['NUTS2'].unique()

array(['IE04', 'IE05', 'ES11', 'PT16', 'PT11', 'PT18', 'UKN0', 'IE06',
       'UKM6', 'ES43', 'ES61', 'ES41', 'UKM9', 'UKL1', 'UKK3', 'ES42',
       'UKM7', 'UKK4', 'FRH0', 'ES13', 'ES30', 'UKD1', 'UKL2', 'UKM5',
       'UKC2', 'UKD3', 'UKG2', 'UKG1', 'UKK2', 'ES21', 'ES23', 'UKE2',
       'UKE3', 'UKK1', 'FRG0', 'ES22', 'ES24', 'ES62', 'UKF2', 'UKJ1',
       'UKJ3', 'FRD1', 'FRI3', 'FRI1', 'ES52', 'UKF3', 'UKH2', 'UKJ2',
       'FRJ2', 'UKH1', 'UKH3', 'UKJ4', 'FRD2', 'FRB0', 'ES51', 'FRI2',
       'FRE1', 'FRE2', 'FR10', 'FRJ1', 'FRK1', 'NL', 'FRC1', 'BE23',
       'BE32', 'FRF2', 'FRK2', 'NO0A', 'BE21', 'BE35', 'FRL0', 'BE33',
       'BE34', 'FRF3', 'FRC2', 'DEA1', 'DEA2', 'LU00', 'CH01', 'NO09',
       'DE94', 'DEA3', 'DEB1', 'FRF1', 'CH02', 'ITC1', 'NO02', 'NO08',
       'DEA5', 'DE72', 'DEB3', 'DE13', 'CH03', 'DK04', 'DK03', 'DE93',
       'DE92', 'DEA4', 'DE73', 'DE71', 'DE12', 'DE14', 'CH04', 'CH06',
       'ITC4', 'ITC3', 'FRM0', 'ITG2', 'NO06', 'DEF0', 'DE11', 'CH05',
       '

# Checkpoint 1: Reading full weather data to avoid data overload

In [1]:
# import needed libraries again and load data obtained from the steps before
import pandas as pd
import numpy as np
import geopandas as gpd
import matplotlib.pyplot as plt
#final_df = pd.read_csv('Heat_wave.csv')
final_df.tail()

lat_long        date  longitude  latitude  wind_2AM  \
3891091  ('61.75', '29.75')  2021-09-26      29.75     61.75      2.28   
3891092  ('61.75', '29.75')  2021-09-27      29.75     61.75     -0.01   
3891093  ('61.75', '29.75')  2021-09-28      29.75     61.75     -1.48   
3891094  ('61.75', '29.75')  2021-09-29      29.75     61.75     -1.20   
3891095  ('61.75', '29.75')  2021-09-30      29.75     61.75     -1.08   

         dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
3891091           5.96             6.96      0.0                      4.81   
3891092           2.83             3.14      0.0                      0.60   
3891093           3.60             3.85      0.0                      1.37   
3891094           5.68             6.56      0.0                      4.36   
3891095           5.78             6.76      0.0                      4.58   

         wind_11AM  ...  MRT_4PM  MRT_11PM  Week_Number  99th_percentile  day  \
3891091       1.17  ...     6.60      0.10           38        26.917881   26   
3891092       0.18  ...     3.04      0.14           39        26.917881   27   
3891093      -1.15  ...     6.69      5.08           39        26.917881   28   
3891094      -0.11  ...     6.52      5.48           39        26.917881   29   
3891095      -0.67  ...     6.30      3.14           39        26.917881   30   

         month  year  threshold_exceeded  consecutive_HW_days  heat_wave_day  
3891091      9  2021                   0                    0              0  
3891092      9  2021                   0                    0              0  
3891093      9  2021                   0                    0              0  
3891094      9  2021                   0                    0              0  
3891095      9  2021                   0                    0              0  

[5 rows x 40 columns]

In [2]:
# inspect shape
final_df.shape

(3891096, 40)

# Merging with health expenditure and disease prevalence data

In [3]:
# read health expenditure csv
expend = pd.read_csv('health_expenditure.csv', index_col=0)
expend.head(2)

country  year  health_expend_p_capita
0      BE  2000             2176.507157
1      BG  2000              270.016329

In [4]:
# merge health expenditure with weather data on country and year
final_df = final_df.merge(expend, how='left', on=['country', 'year'])#.drop(['TIME_y'], axis = 1)
final_df.head(2)

lat_long        date  longitude  latitude  wind_2AM  \
0  ('54.25', '-10.0')  2000-05-01      -10.0     54.25     -3.17   
1  ('54.25', '-10.0')  2000-05-02      -10.0     54.25     -1.20   

   dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
0           7.57            10.66    -0.02                      8.82   
1           8.22             9.81    -0.02                      8.13   

   wind_11AM  ...  MRT_11PM  Week_Number  99th_percentile  day  month  year  \
0       0.16  ...      3.84           18        20.898722    1      5  2000   
1      -1.53  ...      5.60           18        20.898722    2      5  2000   

   threshold_exceeded  consecutive_HW_days  heat_wave_day  \
0                   0                    0              0   
1                   0                    0              0   

   health_expend_p_capita  
0             3760.622022  
1             3760.622022  

[2 rows x 41 columns]

In [5]:
# read disease prevalence csv
prevalence = pd.read_csv('prevalence.csv', index_col=0)
prevalence = prevalence.rename(columns={'Unnamed: 10': 'country'})
prevalence.index.names = ['c']
prevalence.head(2)

Asthma  Chronic_depression  Chronic_lower_respiratory_diseases  \
c                                                                   
0    4.35                 7.6                                 4.4   
1    5.05                 7.0                                 4.0   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
c                                                        
0                                        2.7      5.45   
1                                        1.5      5.55   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
c                                                                              
0                                               1.35                   21.45   
1                                               0.90                   16.95   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke country  
c                                                                              
0             1.95     15.7                                      1.15      AT  
1             0.95     15.0                                      0.90      BE

In [6]:
# merge on country
final_df = final_df.merge(prevalence, how='left', on=['country'])#.drop(['TIME_y'], axis = 1)
final_df.head(2)

lat_long        date  longitude  latitude  wind_2AM  \
0  ('54.25', '-10.0')  2000-05-01      -10.0     54.25     -3.17   
1  ('54.25', '-10.0')  2000-05-02      -10.0     54.25     -1.20   

   dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
0           7.57            10.66    -0.02                      8.82   
1           8.22             9.81    -0.02                      8.13   

   wind_11AM  ...  Asthma  Chronic_depression  \
0       0.16  ...    8.15                 8.0   
1      -1.53  ...    8.15                 8.0   

   Chronic_lower_respiratory_diseases  \
0                                 2.5   
1                                 2.5   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                        2.4      4.05   
1                                        2.4      4.05   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                                1.2                   13.55   
1                                                1.2                   13.55   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  
0              2.8    22.05                                       0.7  
1              2.8    22.05                                       0.7  

[2 rows x 51 columns]

# Merging with mortality data

In [8]:
# read mortality data
mortality = pd.read_csv('Mortality1.csv', index_col=0)
mortality.head()

TIME  GEO                                          GEO_LABEL  deaths  \
17  2000W18  BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   196.0   
18  2000W19  BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   172.0   
19  2000W20  BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   158.0   
20  2000W21  BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   193.0   
21  2000W22  BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   170.0   

   country week_nr  year  population  density  population_65+  share_over_65  \
17      BE     W18  2000    959318.0   5974.3        160908.0      16.773166   
18      BE     W19  2000    959318.0   5974.3        160908.0      16.773166   
19      BE     W20  2000    959318.0   5974.3        160908.0      16.773166   
20      BE     W21  2000    959318.0   5974.3        160908.0      16.773166   
21      BE     W22  2000    959318.0   5974.3        160908.0      16.773166   

    death_p_100k  code_length  
17     20.431181          3.0  
18     17.929404          3.0  
19     16.470034          3.0  
20     20.118459          3.0  
21     17.720923          3.0

In [9]:
# describe deaths per 100,000
mortality['death_p_100k'].describe()

count    129597.000000
mean         18.561734
std           4.409888
min           0.000000
25%          15.789557
50%          18.111807
75%          20.923414
max          56.800635
Name: death_p_100k, dtype: float64

In [10]:
# inspect shape
mortality.shape

(167486, 13)

In [11]:
# drop unnecessary column and rename others
mortality = mortality.drop('week_nr', 1)
mortality = mortality.rename(columns={'TIME': 'week_nr', 'GEO': 'NUTS2'})
mortality.head()

week_nr NUTS2                                          GEO_LABEL  deaths  \
17  2000W18   BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   196.0   
18  2000W19   BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   172.0   
19  2000W20   BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   158.0   
20  2000W21   BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   193.0   
21  2000W22   BE1  Région de Bruxelles-Capitale/Brussels Hoofdste...   170.0   

   country  year  population  density  population_65+  share_over_65  \
17      BE  2000    959318.0   5974.3        160908.0      16.773166   
18      BE  2000    959318.0   5974.3        160908.0      16.773166   
19      BE  2000    959318.0   5974.3        160908.0      16.773166   
20      BE  2000    959318.0   5974.3        160908.0      16.773166   
21      BE  2000    959318.0   5974.3        160908.0      16.773166   

    death_p_100k  code_length  
17     20.431181          3.0  
18     17.929404          3.0  
19     16.470034          3.0  
20     20.118459          3.0  
21     17.720923          3.0

In [12]:
# change and create new variables that match the variables in the mortality data for merging
final_df['year'] = final_df['year'].astype(str)
final_df['Week_Number'] = final_df['Week_Number'].astype(str)
final_df['Week_Number'] = 'W' + final_df['Week_Number']
final_df["week_nr"] = final_df['year'] + final_df['Week_Number']
final_df.head()

lat_long        date  longitude  latitude  wind_2AM  \
0  ('54.25', '-10.0')  2000-05-01      -10.0     54.25     -3.17   
1  ('54.25', '-10.0')  2000-05-02      -10.0     54.25     -1.20   
2  ('54.25', '-10.0')  2000-05-03      -10.0     54.25     -4.29   
3  ('54.25', '-10.0')  2000-05-04      -10.0     54.25     -4.35   
4  ('54.25', '-10.0')  2000-05-05      -10.0     54.25     -3.99   

   dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
0           7.57            10.66    -0.02                      8.82   
1           8.22             9.81    -0.02                      8.13   
2           6.97             8.52    -0.02                      6.56   
3           6.38             8.56    -0.02                      6.47   
4           7.33             9.03    -0.02                      7.15   

   wind_11AM  ...  Chronic_depression  Chronic_lower_respiratory_diseases  \
0       0.16  ...                 8.0                                 2.5   
1      -1.53  ...                 8.0                                 2.5   
2      -4.15  ...                 8.0                                 2.5   
3      -4.60  ...                 8.0                                 2.5   
4      -3.51  ...                 8.0                                 2.5   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                        2.4      4.05   
1                                        2.4      4.05   
2                                        2.4      4.05   
3                                        2.4      4.05   
4                                        2.4      4.05   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                                1.2                   13.55   
1                                                1.2                   13.55   
2                                                1.2                   13.55   
3                                                1.2                   13.55   
4                                                1.2                   13.55   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  week_nr  
0              2.8    22.05                                       0.7  2000W18  
1              2.8    22.05                                       0.7  2000W18  
2              2.8    22.05                                       0.7  2000W18  
3              2.8    22.05                                       0.7  2000W18  
4              2.8    22.05                                       0.7  2000W18  

[5 rows x 52 columns]

In [13]:
# inspect shape
final_df.shape

(3891096, 52)

# Merge weather data with mortality

In [14]:
# merge the weather data with mortality on week number and NUTS2
final_df = final_df.merge(mortality, how='left', on=['NUTS2','week_nr']).drop(['country_y', 'year_y'], axis = 1)
final_df = final_df.rename(columns={'GEO_LABEL': 'NUTS2_Label', 'country_x': 'country', 'year_x': 'year'})
final_df.shape

(3891096, 60)

In [15]:
# inspect data
pd.set_option('display.max_columns', None)
final_df.head(2)

lat_long        date  longitude  latitude  wind_2AM  \
0  ('54.25', '-10.0')  2000-05-01      -10.0     54.25     -3.17   
1  ('54.25', '-10.0')  2000-05-02      -10.0     54.25     -1.20   

   dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
0           7.57            10.66    -0.02                      8.82   
1           8.22             9.81    -0.02                      8.13   

   wind_11AM  dew_point_11AM  temperature_11AM   uvb_11AM  \
0       0.16            8.74             13.70  247851.27   
1      -1.53            9.74             11.66  244230.16   

   apparent_temperature_11AM  wind_4PM  dew_point_4PM  temperature_4PM  \
0                      12.13     -1.44           8.74            14.12   
1                      10.39     -2.43           8.64            11.02   

     uvb_4PM  apparent_temperature_4PM  wind_11PM  dew_point_11PM  \
0  232877.31                     12.55      -1.94            8.07   
1  180512.10                      9.44      -3.55            7.47   

   temperature_11PM  uvb_11PM  apparent_temperature_11PM country NUTS1 NUTS2  \
0             10.32     -0.02                       8.60      IE   IE0  IE04   
1              8.99     -0.02                       7.14      IE   IE0  IE04   

   NUTS3  MRT_2AM  MRT_11AM  MRT_4PM  MRT_11PM Week_Number  99th_percentile  \
0  IE042     2.96     23.10    15.57      3.84         W18        20.898722   
1  IE042     3.03     23.49    25.09      5.60         W18        20.898722   

   day  month  year  threshold_exceeded  consecutive_HW_days  heat_wave_day  \
0    1      5  2000                   0                    0              0   
1    2      5  2000                   0                    0              0   

   health_expend_p_capita  Asthma  Chronic_depression  \
0             3760.622022    8.15                 8.0   
1             3760.622022    8.15                 8.0   

   Chronic_lower_respiratory_diseases  \
0                                 2.5   
1                                 2.5   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                        2.4      4.05   
1                                        2.4      4.05   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                                1.2                   13.55   
1                                                1.2                   13.55   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  \
0              2.8    22.05                                       0.7   
1              2.8    22.05                                       0.7   

   week_nr NUTS2_Label  deaths  population  density  population_65+  \
0  2000W18         NaN     NaN         NaN      NaN             NaN   
1  2000W18         NaN     NaN         NaN      NaN             NaN   

   share_over_65  death_p_100k  code_length  
0            NaN           NaN          NaN  
1            NaN           NaN          NaN

## Dropping days with NaN for mortality

In [21]:
# dropping observations with NaN for mortality
final_df = final_df.dropna(subset=['death_p_100k'])
final_df.shape

(2550190, 60)

The number of rows with available mortality data equals 2,555,190, down from 3,891,096

# Checkpoint 2: Writing dataframe without NaN for mortality to CSV

In [2]:
# 2nd checkpoint aimed to prevent data overload
# read and load data
#final_df.to_csv('data_no_NaN.csv')
final_df = pd.read_csv('data_no_NaN.csv')
final_df = final_df.iloc[: , 1:]

In [3]:
# inspect data
final_df.head()

lat_long        date  longitude  latitude  wind_2AM  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0     -2.12   
1  ('43.0', '-9.25')  2000-05-02      -9.25      43.0     -1.48   
2  ('43.0', '-9.25')  2000-05-03      -9.25      43.0     -3.79   
3  ('43.0', '-9.25')  2000-05-04      -9.25      43.0     -1.34   
4  ('43.0', '-9.25')  2000-05-05      -9.25      43.0     -2.68   

   dew_point_2AM  temperature_2AM  uvb_2AM  apparent_temperature_2AM  \
0          11.14            12.81    -0.02                     11.98   
1          12.80            13.52    -0.02                     13.29   
2          12.36            13.44    -0.02                     13.05   
3          13.01            14.00    -0.02                     13.86   
4          12.54            13.75    -0.02                     13.42   

   wind_11AM  ...  Stroke_or_chronic_consequences_of_stroke  week_nr  \
0      -3.62  ...                                       0.9  2000W18   
1       1.25  ...                                       0.9  2000W18   
2      -4.98  ...                                       0.9  2000W18   
3       0.34  ...                                       0.9  2000W18   
4       0.05  ...                                       0.9  2000W18   

   NUTS2_Label  deaths  population  density  population_65+  share_over_65  \
0      Galicia   541.0   2702471.0     91.3        544528.0      20.149263   
1      Galicia   541.0   2702471.0     91.3        544528.0      20.149263   
2      Galicia   541.0   2702471.0     91.3        544528.0      20.149263   
3      Galicia   541.0   2702471.0     91.3        544528.0      20.149263   
4      Galicia   541.0   2702471.0     91.3        544528.0      20.149263   

   death_p_100k  code_length  
0     20.018716          NaN  
1     20.018716          NaN  
2     20.018716          NaN  
3     20.018716          NaN  
4     20.018716          NaN  

[5 rows x 60 columns]

In [4]:
# # To reducte the number of features, wind and dew_point will be kept as a daily average instead of hourly

# wind
final_df['wind_2AM'] = abs(final_df['wind_2AM'])
final_df['wind_11AM'] = abs(final_df['wind_11AM'])
final_df['wind_4PM'] = abs(final_df['wind_4PM'])
final_df['wind_11PM'] = abs(final_df['wind_11PM'])
final_df['wind_day'] = (final_df['wind_2AM'] + final_df['wind_11AM'] + final_df['wind_4PM'] + final_df['wind_11PM']) / 4
final_df = final_df.drop(['wind_2AM', 'wind_11AM', 'wind_4PM', 'wind_11PM'],1)

# dew_point
final_df['dew_point_2AM'] = final_df['dew_point_2AM']
final_df['dew_point_11AM'] = final_df['dew_point_11AM']
final_df['dew_point_4PM'] = final_df['dew_point_4PM']
final_df['dew_point_11PM'] = final_df['dew_point_11PM']
final_df['dew_point_day'] = (final_df['dew_point_2AM'] + final_df['dew_point_11AM'] + final_df['dew_point_4PM'] + final_df['dew_point_11PM']) / 4
final_df = final_df.drop(['dew_point_2AM', 'dew_point_11AM', 'dew_point_4PM', 'dew_point_11PM'],1)


In [5]:
# uvb radiation measures during the evening, night and before noon are considered irrelevant and dropped
final_df = final_df.drop(['uvb_2AM', 'uvb_11AM', 'uvb_11PM'],1)

## Computing three-year rolling average in mortality

In [6]:
# for each location, compute the yearly mean
avg_demo = pd.DataFrame(final_df.groupby(['lat_long', 'year'])['death_p_100k'].mean())
avg_demo = avg_demo.add_suffix('_avg')
avg_demo = avg_demo.reset_index(level=['lat_long', 'year'])
avg_demo = avg_demo.rename(columns={'NUTS1_avg': 'NUTS1'})
#df.groupby('A').mean()
avg_demo.head()

lat_long  year  death_p_100k_avg
0  ('37.0', '-2.5')  2000         14.380536
1  ('37.0', '-2.5')  2001         15.033509
2  ('37.0', '-2.5')  2002         14.396432
3  ('37.0', '-2.5')  2003         15.871519
4  ('37.0', '-2.5')  2004         14.475298

In [7]:
avg_demo.shape

(16704, 3)

In [8]:
# get three-year rolling average of weekly mortality centered at year per locations
avg_demo['death_p_100k_3_avg'] = avg_demo.groupby('lat_long')['death_p_100k_avg'].rolling(window=3, center=True).mean().reset_index(0,drop=True)
avg_demo.head(25)

lat_long  year  death_p_100k_avg  death_p_100k_3_avg
0    ('37.0', '-2.5')  2000         14.380536                 NaN
1    ('37.0', '-2.5')  2001         15.033509           14.603492
2    ('37.0', '-2.5')  2002         14.396432           15.100487
3    ('37.0', '-2.5')  2003         15.871519           14.914416
4    ('37.0', '-2.5')  2004         14.475298           14.755544
5    ('37.0', '-2.5')  2005         13.919816           14.247227
6    ('37.0', '-2.5')  2006         14.346568           14.144199
7    ('37.0', '-2.5')  2007         14.166212           14.198117
8    ('37.0', '-2.5')  2008         14.081571           13.975246
9    ('37.0', '-2.5')  2009         13.677954           13.889492
10   ('37.0', '-2.5')  2010         13.908952           13.804776
11   ('37.0', '-2.5')  2011         13.827422           13.781223
12   ('37.0', '-2.5')  2012         13.607297           13.751694
13   ('37.0', '-2.5')  2013         13.820364           13.738614
14   ('37.0', '-2.5')  2014         13.788180           14.041433
15   ('37.0', '-2.5')  2015         14.515755           14.341510
16   ('37.0', '-2.5')  2016         14.720595           14.565851
17   ('37.0', '-2.5')  2017         14.461204           14.582050
18   ('37.0', '-2.5')  2018         14.564352           14.513948
19   ('37.0', '-2.5')  2019         14.516288           14.877026
20   ('37.0', '-2.5')  2020         15.550439           15.599485
21   ('37.0', '-2.5')  2021         16.731728                 NaN
22  ('37.0', '-3.25')  2000         14.380536                 NaN
23  ('37.0', '-3.25')  2001         15.033509           14.603492
24  ('37.0', '-3.25')  2002         14.396432           15.100487

Missing values for 2021 and 2000 as no average centered at these year can be computed due to data from 1999 and 2022 not being available.

In [9]:
# impute NaN of years 2000 with the average centered at 2001
df_21 = avg_demo[avg_demo['year'] == 2021]
avg_demo = avg_demo[avg_demo['year'] != 2021]
avg_demo['death_p_100k_3_avg'].fillna(method='bfill', inplace=True)
avg_demo.head(5)

lat_long  year  death_p_100k_avg  death_p_100k_3_avg
0  ('37.0', '-2.5')  2000         14.380536           14.603492
1  ('37.0', '-2.5')  2001         15.033509           14.603492
2  ('37.0', '-2.5')  2002         14.396432           15.100487
3  ('37.0', '-2.5')  2003         15.871519           14.914416
4  ('37.0', '-2.5')  2004         14.475298           14.755544

In [10]:
# impute NaN for 2021 with the averages for only 2021 to account for higher death rates because of Covid
df_21['death_p_100k_3_avg'] = df_21['death_p_100k_avg']
frames = [avg_demo,df_21]
avg_demo = pd.concat(frames)
avg_demo.sort_values(by=['lat_long', 'year'], inplace=True)
avg_demo.head(25)

lat_long  year  death_p_100k_avg  death_p_100k_3_avg
0    ('37.0', '-2.5')  2000         14.380536           14.603492
1    ('37.0', '-2.5')  2001         15.033509           14.603492
2    ('37.0', '-2.5')  2002         14.396432           15.100487
3    ('37.0', '-2.5')  2003         15.871519           14.914416
4    ('37.0', '-2.5')  2004         14.475298           14.755544
5    ('37.0', '-2.5')  2005         13.919816           14.247227
6    ('37.0', '-2.5')  2006         14.346568           14.144199
7    ('37.0', '-2.5')  2007         14.166212           14.198117
8    ('37.0', '-2.5')  2008         14.081571           13.975246
9    ('37.0', '-2.5')  2009         13.677954           13.889492
10   ('37.0', '-2.5')  2010         13.908952           13.804776
11   ('37.0', '-2.5')  2011         13.827422           13.781223
12   ('37.0', '-2.5')  2012         13.607297           13.751694
13   ('37.0', '-2.5')  2013         13.820364           13.738614
14   ('37.0', '-2.5')  2014         13.788180           14.041433
15   ('37.0', '-2.5')  2015         14.515755           14.341510
16   ('37.0', '-2.5')  2016         14.720595           14.565851
17   ('37.0', '-2.5')  2017         14.461204           14.582050
18   ('37.0', '-2.5')  2018         14.564352           14.513948
19   ('37.0', '-2.5')  2019         14.516288           14.877026
20   ('37.0', '-2.5')  2020         15.550439           15.599485
21   ('37.0', '-2.5')  2021         16.731728           16.731728
22  ('37.0', '-3.25')  2000         14.380536           14.603492
23  ('37.0', '-3.25')  2001         15.033509           14.603492
24  ('37.0', '-3.25')  2002         14.396432           15.100487

In [11]:
# keep only the three-year averages for each year for each location
avg_demo = avg_demo.drop(['death_p_100k_avg'],1)
avg_demo = avg_demo.rename(columns={'death_p_100k_3_avg': 'death_p_100k_avg'})
avg_demo.head(1)

lat_long  year  death_p_100k_avg
0  ('37.0', '-2.5')  2000         14.603492

In [12]:
# write data to csv
avg_demo.to_csv('avg_deaths_100k.csv')

## Get minimum and maximum temperature

In [13]:
temperatures = final_df[['apparent_temperature_2AM', 'apparent_temperature_11AM', 'apparent_temperature_4PM', 'apparent_temperature_11PM']]
# find the maximum and minimum values of each row
temperatures['max_apparent_temperature'] = temperatures.max(axis = 1)
temperatures['min_apparent_temperature'] = temperatures.min(axis = 1)
temperatures = temperatures[['max_apparent_temperature', 'min_apparent_temperature']]
temperatures

<ipython-input-13-04c9feebc24b>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temperatures['max_apparent_temperature'] = temperatures.max(axis = 1)
<ipython-input-13-04c9feebc24b>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temperatures['min_apparent_temperature'] = temperatures.min(axis = 1)


max_apparent_temperature  min_apparent_temperature
0                           15.87                     11.98
1                           15.83                     12.95
2                           18.53                     13.05
3                           15.94                     13.86
4                           16.34                     13.42
...                           ...                       ...
2550185                      8.36                      1.22
2550186                      9.65                      0.60
2550187                      6.55                      1.37
2550188                      6.85                      4.36
2550189                      6.58                      4.14

[2550190 rows x 2 columns]

In [14]:
# drop hourly apparent temperature features and other ones that are deemed irrelevant for the modelling
final_df = final_df.drop(['apparent_temperature_2AM', 'apparent_temperature_11AM', 'apparent_temperature_4PM', 
                          'apparent_temperature_11PM', 'temperature_2AM', 'temperature_4PM', 'MRT_11AM', 'MRT_11PM',
                          'temperature_11AM','temperature_11PM'],1)
# join with obtained max and minimum apparent temperature 
final_df = final_df.join(temperatures)
final_df

lat_long        date  longitude  latitude    uvb_4PM  \
0         ('43.0', '-9.25')  2000-05-01      -9.25     43.00  165578.40   
1         ('43.0', '-9.25')  2000-05-02      -9.25     43.00  164847.47   
2         ('43.0', '-9.25')  2000-05-03      -9.25     43.00  244458.16   
3         ('43.0', '-9.25')  2000-05-04      -9.25     43.00  150456.94   
4         ('43.0', '-9.25')  2000-05-05      -9.25     43.00  100901.43   
...                     ...         ...        ...       ...        ...   
2550185  ('61.75', '29.75')  2021-09-26      29.75     61.75    4301.75   
2550186  ('61.75', '29.75')  2021-09-27      29.75     61.75    4378.33   
2550187  ('61.75', '29.75')  2021-09-28      29.75     61.75    2498.92   
2550188  ('61.75', '29.75')  2021-09-29      29.75     61.75    2234.41   
2550189  ('61.75', '29.75')  2021-09-30      29.75     61.75    1531.38   

        country NUTS1 NUTS2  NUTS3  MRT_2AM  ...  population density  \
0            ES   ES1  ES11  ES111     6.35  ...   2702471.0    91.3   
1            ES   ES1  ES11  ES111     8.06  ...   2702471.0    91.3   
2            ES   ES1  ES11  ES111     5.84  ...   2702471.0    91.3   
3            ES   ES1  ES11  ES111     9.93  ...   2702471.0    91.3   
4            ES   ES1  ES11  ES111     6.88  ...   2702471.0    91.3   
...         ...   ...   ...    ...      ...  ...         ...     ...   
2550185      FI   FI1  FI1C  FI1C5     4.13  ...   1148794.0    36.6   
2550186      FI   FI1  FI1C  FI1C5     0.34  ...   1148794.0    36.6   
2550187      FI   FI1  FI1C  FI1C5     3.95  ...   1148794.0    36.6   
2550188      FI   FI1  FI1C  FI1C5     4.56  ...   1148794.0    36.6   
2550189      FI   FI1  FI1C  FI1C5     4.37  ...   1148794.0    36.6   

         population_65+  share_over_65  death_p_100k  code_length  wind_day  \
0              544528.0      20.149263     20.018716          NaN    2.2075   
1              544528.0      20.149263     20.018716          NaN    1.7075   
2              544528.0      20.149263     20.018716          NaN    3.4325   
3              544528.0      20.149263     20.018716          NaN    1.2650   
4              544528.0      20.149263     20.018716          NaN    1.2750   
...                 ...            ...           ...          ...       ...   
2550185        291137.0      25.342838     24.982721          NaN    1.4200   
2550186        291137.0      25.342838     23.067669          NaN    0.5250   
2550187        291137.0      25.342838     23.067669          NaN    1.1925   
2550188        291137.0      25.342838     23.067669          NaN    1.0125   
2550189        291137.0      25.342838     23.067669          NaN    1.1600   

         dew_point_day  max_apparent_temperature  min_apparent_temperature  
0              12.3350                     15.87                     11.98  
1              12.9650                     15.83                     12.95  
2              12.8600                     18.53                     13.05  
3              13.1650                     15.94                     13.86  
4              12.7725                     16.34                     13.42  
...                ...                       ...                       ...  
2550185         5.4525                      8.36                      1.22  
2550186         4.5175                      9.65                      0.60  
2550187         5.4500                      6.55                      1.37  
2550188         5.2575                      6.85                      4.36  
2550189         4.1350                      6.58                      4.14  

[2550190 rows x 43 columns]

### Compute averages of weather variables per location as the deviation from the average per location might be an informative feature

In [15]:
# get average of weather variables per locations
avg = pd.DataFrame(final_df.groupby('lat_long')['wind_day', 'dew_point_day', 'uvb_4PM','max_apparent_temperature', 'min_apparent_temperature',
                                                'MRT_2AM', 'MRT_4PM'].mean())
avg = avg.add_suffix('_avg')
avg = avg.reset_index(level=['lat_long'])
avg = avg.rename(columns={'NUTS1_avg': 'NUTS1'})
#df.groupby('A').mean()
avg.head()

<ipython-input-15-d27a906214a1>:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  avg = pd.DataFrame(final_df.groupby('lat_long')['wind_day', 'dew_point_day', 'uvb_4PM','max_apparent_temperature', 'min_apparent_temperature',


lat_long  wind_day_avg  dew_point_day_avg    uvb_4PM_avg  \
0   ('37.0', '-2.5')      1.359658          11.427329  230179.699985   
1  ('37.0', '-3.25')      0.940888           8.217627  232642.714719   
2   ('37.0', '-4.0')      0.852583          12.090778  238715.081325   
3  ('37.0', '-4.75')      1.172296          12.823071  243280.345409   
4   ('37.0', '-5.5')      1.696267          12.777655  247985.496707   

   max_apparent_temperature_avg  min_apparent_temperature_avg  MRT_2AM_avg  \
0                     27.170423                     17.870845    10.191060   
1                     23.415918                     13.653343     6.793954   
2                     26.272641                     16.184790     8.756076   
3                     28.053876                     16.880438     8.883504   
4                     30.056937                     18.812245    10.981810   

   MRT_4PM_avg  
0    35.969705  
1    32.017990  
2    34.119667  
3    34.742236  
4    37.806311

In [16]:
# store columns that can be compared to the average in a list
b = final_df.columns.values
b = b.tolist()
b = [e for e in b if e not in ('lat_long', 'date', 'longitude', 'latitude','Week_Number', '99th_percentile', 'day', 'month', 'year',
         'threshold_exceeded', 'consecutive_HW_days', 'heat_wave_day','week_nr', 'NUTS1_Label', 'deaths', 'population', 'population_65+', 'share_over_65', 'density',
         'country', 'NUTS1', 'NUTS2', 'NUTS3', 'health_expend_p_capita', 'Asthma', 'Chronic_depression', 'Chronic_lower_respiratory_diseases',
                               'Coronary_heart_disease_or_angina_pectoris', 'Diabetes', 'Heart_attack_or_chronic_consequences_of_heart_attack', 
                               'High_blood_pressure', 'Kidney_problems', 'Stroke_or_chronic_consequences_of_stroke', 'Obesity', 'NUTS2_Label','code_length')]

print(b)

['uvb_4PM', 'MRT_2AM', 'MRT_4PM', 'death_p_100k', 'wind_day', 'dew_point_day', 'max_apparent_temperature', 'min_apparent_temperature']


In [17]:
# adding averages per location to dataframe
final_df = final_df.merge(avg, how='inner', on=['lat_long'])
final_df.head()

lat_long        date  longitude  latitude    uvb_4PM country  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0  165578.40      ES   
1  ('43.0', '-9.25')  2000-05-02      -9.25      43.0  164847.47      ES   
2  ('43.0', '-9.25')  2000-05-03      -9.25      43.0  244458.16      ES   
3  ('43.0', '-9.25')  2000-05-04      -9.25      43.0  150456.94      ES   
4  ('43.0', '-9.25')  2000-05-05      -9.25      43.0  100901.43      ES   

  NUTS1 NUTS2  NUTS3  MRT_2AM  ...  dew_point_day max_apparent_temperature  \
0   ES1  ES11  ES111     6.35  ...        12.3350                    15.87   
1   ES1  ES11  ES111     8.06  ...        12.9650                    15.83   
2   ES1  ES11  ES111     5.84  ...        12.8600                    18.53   
3   ES1  ES11  ES111     9.93  ...        13.1650                    15.94   
4   ES1  ES11  ES111     6.88  ...        12.7725                    16.34   

   min_apparent_temperature  wind_day_avg  dew_point_day_avg    uvb_4PM_avg  \
0                     11.98      2.809025          13.979091  225349.668363   
1                     12.95      2.809025          13.979091  225349.668363   
2                     13.05      2.809025          13.979091  225349.668363   
3                     13.86      2.809025          13.979091  225349.668363   
4                     13.42      2.809025          13.979091  225349.668363   

   max_apparent_temperature_avg  min_apparent_temperature_avg  MRT_2AM_avg  \
0                     18.963855                      15.72365    10.002745   
1                     18.963855                      15.72365    10.002745   
2                     18.963855                      15.72365    10.002745   
3                     18.963855                      15.72365    10.002745   
4                     18.963855                      15.72365    10.002745   

   MRT_4PM_avg  
0      24.7651  
1      24.7651  
2      24.7651  
3      24.7651  
4      24.7651  

[5 rows x 50 columns]

In [18]:
# merge final_df with average mortality per year and location
final_df = final_df.merge(avg_demo, how='inner', on=['lat_long', 'year'])
final_df.head()

lat_long        date  longitude  latitude    uvb_4PM country  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0  165578.40      ES   
1  ('43.0', '-9.25')  2000-05-02      -9.25      43.0  164847.47      ES   
2  ('43.0', '-9.25')  2000-05-03      -9.25      43.0  244458.16      ES   
3  ('43.0', '-9.25')  2000-05-04      -9.25      43.0  150456.94      ES   
4  ('43.0', '-9.25')  2000-05-05      -9.25      43.0  100901.43      ES   

  NUTS1 NUTS2  NUTS3  MRT_2AM  ...  max_apparent_temperature  \
0   ES1  ES11  ES111     6.35  ...                     15.87   
1   ES1  ES11  ES111     8.06  ...                     15.83   
2   ES1  ES11  ES111     5.84  ...                     18.53   
3   ES1  ES11  ES111     9.93  ...                     15.94   
4   ES1  ES11  ES111     6.88  ...                     16.34   

  min_apparent_temperature  wind_day_avg  dew_point_day_avg    uvb_4PM_avg  \
0                    11.98      2.809025          13.979091  225349.668363   
1                    12.95      2.809025          13.979091  225349.668363   
2                    13.05      2.809025          13.979091  225349.668363   
3                    13.86      2.809025          13.979091  225349.668363   
4                    13.42      2.809025          13.979091  225349.668363   

   max_apparent_temperature_avg  min_apparent_temperature_avg  MRT_2AM_avg  \
0                     18.963855                      15.72365    10.002745   
1                     18.963855                      15.72365    10.002745   
2                     18.963855                      15.72365    10.002745   
3                     18.963855                      15.72365    10.002745   
4                     18.963855                      15.72365    10.002745   

   MRT_4PM_avg  death_p_100k_avg  
0      24.7651         18.451122  
1      24.7651         18.451122  
2      24.7651         18.451122  
3      24.7651         18.451122  
4      24.7651         18.451122  

[5 rows x 51 columns]

In [19]:
# compute the deviation of each value from the average value for the location (done for all weather and mortality variables)
pd.set_option('display.max_columns', None)
for i in b:                      
    final_df['{}_vs_avg'.format(i)] = (final_df[i] / final_df['{}_avg'.format(i)])
final_df.head() 

lat_long        date  longitude  latitude    uvb_4PM country  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0  165578.40      ES   
1  ('43.0', '-9.25')  2000-05-02      -9.25      43.0  164847.47      ES   
2  ('43.0', '-9.25')  2000-05-03      -9.25      43.0  244458.16      ES   
3  ('43.0', '-9.25')  2000-05-04      -9.25      43.0  150456.94      ES   
4  ('43.0', '-9.25')  2000-05-05      -9.25      43.0  100901.43      ES   

  NUTS1 NUTS2  NUTS3  MRT_2AM  MRT_4PM Week_Number  99th_percentile  day  \
0   ES1  ES11  ES111     6.35    28.64         W18        25.658925    1   
1   ES1  ES11  ES111     8.06    24.69         W18        25.658925    2   
2   ES1  ES11  ES111     5.84    23.74         W18        25.658925    3   
3   ES1  ES11  ES111     9.93    27.42         W18        25.658925    4   
4   ES1  ES11  ES111     6.88    26.17         W18        25.658925    5   

   month  year  threshold_exceeded  consecutive_HW_days  heat_wave_day  \
0      5  2000                   0                    0              0   
1      5  2000                   0                    0              0   
2      5  2000                   0                    0              0   
3      5  2000                   0                    0              0   
4      5  2000                   0                    0              0   

   health_expend_p_capita  Asthma  Chronic_depression  \
0             1164.962074     4.3                6.75   
1             1164.962074     4.3                6.75   
2             1164.962074     4.3                6.75   
3             1164.962074     4.3                6.75   
4             1164.962074     4.3                6.75   

   Chronic_lower_respiratory_diseases  \
0                                 3.0   
1                                 3.0   
2                                 3.0   
3                                 3.0   
4                                 3.0   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                       0.75      7.15   
1                                       0.75      7.15   
2                                       0.75      7.15   
3                                       0.75      7.15   
4                                       0.75      7.15   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                               0.85                    19.0   
1                                               0.85                    19.0   
2                                               0.85                    19.0   
3                                               0.85                    19.0   
4                                               0.85                    19.0   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  \
0              2.6     16.1                                       0.9   
1              2.6     16.1                                       0.9   
2              2.6     16.1                                       0.9   
3              2.6     16.1                                       0.9   
4              2.6     16.1                                       0.9   

   week_nr NUTS2_Label  deaths  population  density  population_65+  \
0  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
1  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
2  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
3  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
4  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   

   share_over_65  death_p_100k  code_length  wind_day  dew_point_day  \
0      20.149263     20.018716          NaN    2.2075        12.3350   
1      20.149263     20.018716          NaN    1.7075        12.9650   
2      20.149263     20.018716          NaN    3.4325        12.8600   
3      20.149263     20.018716          NaN    1.2650        13.1650   
4      

The results are new column that indicate the deviation from the mean with suffix 'vs_avg'. If the columns 'vs_avg' = 1, they are equal to the average, if > 1, they are larger than the average.

In [21]:
# combine coordinates with week number in new variable to get a unique identifier for each calendar week at each location
final_df['unique_week'] = final_df['lat_long'] + final_df['week_nr']
final_df.head(1)

lat_long        date  longitude  latitude   uvb_4PM country NUTS1  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0  165578.4      ES   ES1   

  NUTS2  NUTS3  MRT_2AM  MRT_4PM Week_Number  99th_percentile  day  month  \
0  ES11  ES111     6.35    28.64         W18        25.658925    1      5   

   year  threshold_exceeded  consecutive_HW_days  heat_wave_day  \
0  2000                   0                    0              0   

   health_expend_p_capita  Asthma  Chronic_depression  \
0             1164.962074     4.3                6.75   

   Chronic_lower_respiratory_diseases  \
0                                 3.0   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                       0.75      7.15   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                               0.85                    19.0   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  \
0              2.6     16.1                                       0.9   

   week_nr NUTS2_Label  deaths  population  density  population_65+  \
0  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   

   share_over_65  death_p_100k  code_length  wind_day  dew_point_day  \
0      20.149263     20.018716          NaN    2.2075         12.335   

   max_apparent_temperature  min_apparent_temperature  wind_day_avg  \
0                     15.87                     11.98      2.809025   

   dew_point_day_avg    uvb_4PM_avg  max_apparent_temperature_avg  \
0          13.979091  225349.668363                     18.963855   

   min_apparent_temperature_avg  MRT_2AM_avg  MRT_4PM_avg  death_p_100k_avg  \
0                      15.72365    10.002745      24.7651         18.451122   

   uvb_4PM_vs_avg  MRT_2AM_vs_avg  MRT_4PM_vs_avg  death_p_100k_vs_avg  \
0        0.734762        0.634826        1.156466             1.084959   

   wind_day_vs_avg  dew_point_day_vs_avg  max_apparent_temperature_vs_avg  \
0          0.78586              0.882389                         0.836855   

   min_apparent_temperature_vs_avg               unique_week  
0                          0.76191  ('43.0', '-9.25')2000W18

In [22]:
# get the sum of consecutive days of the threshold exceeded starting from the previous instance
final_df['consecutive_HW_days_prev'] = final_df['consecutive_HW_days'].shift()
final_df.head()

lat_long        date  longitude  latitude    uvb_4PM country  \
0  ('43.0', '-9.25')  2000-05-01      -9.25      43.0  165578.40      ES   
1  ('43.0', '-9.25')  2000-05-02      -9.25      43.0  164847.47      ES   
2  ('43.0', '-9.25')  2000-05-03      -9.25      43.0  244458.16      ES   
3  ('43.0', '-9.25')  2000-05-04      -9.25      43.0  150456.94      ES   
4  ('43.0', '-9.25')  2000-05-05      -9.25      43.0  100901.43      ES   

  NUTS1 NUTS2  NUTS3  MRT_2AM  MRT_4PM Week_Number  99th_percentile  day  \
0   ES1  ES11  ES111     6.35    28.64         W18        25.658925    1   
1   ES1  ES11  ES111     8.06    24.69         W18        25.658925    2   
2   ES1  ES11  ES111     5.84    23.74         W18        25.658925    3   
3   ES1  ES11  ES111     9.93    27.42         W18        25.658925    4   
4   ES1  ES11  ES111     6.88    26.17         W18        25.658925    5   

   month  year  threshold_exceeded  consecutive_HW_days  heat_wave_day  \
0      5  2000                   0                    0              0   
1      5  2000                   0                    0              0   
2      5  2000                   0                    0              0   
3      5  2000                   0                    0              0   
4      5  2000                   0                    0              0   

   health_expend_p_capita  Asthma  Chronic_depression  \
0             1164.962074     4.3                6.75   
1             1164.962074     4.3                6.75   
2             1164.962074     4.3                6.75   
3             1164.962074     4.3                6.75   
4             1164.962074     4.3                6.75   

   Chronic_lower_respiratory_diseases  \
0                                 3.0   
1                                 3.0   
2                                 3.0   
3                                 3.0   
4                                 3.0   

   Coronary_heart_disease_or_angina_pectoris  Diabetes  \
0                                       0.75      7.15   
1                                       0.75      7.15   
2                                       0.75      7.15   
3                                       0.75      7.15   
4                                       0.75      7.15   

   Heart_attack_or_chronic_consequences_of_heart_attack  High_blood_pressure  \
0                                               0.85                    19.0   
1                                               0.85                    19.0   
2                                               0.85                    19.0   
3                                               0.85                    19.0   
4                                               0.85                    19.0   

   Kidney_problems  Obesity  Stroke_or_chronic_consequences_of_stroke  \
0              2.6     16.1                                       0.9   
1              2.6     16.1                                       0.9   
2              2.6     16.1                                       0.9   
3              2.6     16.1                                       0.9   
4              2.6     16.1                                       0.9   

   week_nr NUTS2_Label  deaths  population  density  population_65+  \
0  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
1  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
2  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
3  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   
4  2000W18     Galicia   541.0   2702471.0     91.3        544528.0   

   share_over_65  death_p_100k  code_length  wind_day  dew_point_day  \
0      20.149263     20.018716          NaN    2.2075        12.3350   
1      20.149263     20.018716          NaN    1.7075        12.9650   
2      20.149263     20.018716          NaN    3.4325        12.8600   
3      20.149263     20.018716          NaN    1.2650        13.1650   
4      

In [23]:
# get a unique identifier for periods with constant value in 'consecutive_HW_days column'
final_df['cumsum'] = (final_df['consecutive_HW_days'] != final_df['consecutive_HW_days_prev']).cumsum()
final_df

lat_long        date  longitude  latitude    uvb_4PM  \
0        ('43.0', '-9.25')  2000-05-01      -9.25     43.00  165578.40   
1        ('43.0', '-9.25')  2000-05-02      -9.25     43.00  164847.47   
2        ('43.0', '-9.25')  2000-05-03      -9.25     43.00  244458.16   
3        ('43.0', '-9.25')  2000-05-04      -9.25     43.00  150456.94   
4        ('43.0', '-9.25')  2000-05-05      -9.25     43.00  100901.43   
...                    ...         ...        ...       ...        ...   
2550185  ('45.25', '29.0')  2021-09-26      29.00     45.25    9097.72   
2550186  ('45.25', '29.0')  2021-09-27      29.00     45.25   10893.59   
2550187  ('45.25', '29.0')  2021-09-28      29.00     45.25    3932.83   
2550188  ('45.25', '29.0')  2021-09-29      29.00     45.25    7232.23   
2550189  ('45.25', '29.0')  2021-09-30      29.00     45.25    6222.92   

        country NUTS1 NUTS2  NUTS3  MRT_2AM  MRT_4PM Week_Number  \
0            ES   ES1  ES11  ES111     6.35    28.64         W18   
1            ES   ES1  ES11  ES111     8.06    24.69         W18   
2            ES   ES1  ES11  ES111     5.84    23.74         W18   
3            ES   ES1  ES11  ES111     9.93    27.42         W18   
4            ES   ES1  ES11  ES111     6.88    26.17         W18   
...         ...   ...   ...    ...      ...      ...         ...   
2550185      RO   RO2  RO22  RO225     4.35    16.80         W38   
2550186      RO   RO2  RO22  RO225     6.33    10.38         W39   
2550187      RO   RO2  RO22  RO225     5.44    12.95         W39   
2550188      RO   RO2  RO22  RO225     5.20    10.11         W39   
2550189      RO   RO2  RO22  RO225     4.11    10.26         W39   

         99th_percentile  day  month  year  threshold_exceeded  \
0              25.658925    1      5  2000                   0   
1              25.658925    2      5  2000                   0   
2              25.658925    3      5  2000                   0   
3              25.658925    4      5  2000                   0   
4              25.658925    5      5  2000                   0   
...                  ...  ...    ...   ...                 ...   
2550185        33.860439   26      9  2021                   0   
2550186        33.860439   27      9  2021                   0   
2550187        33.860439   28      9  2021                   0   
2550188        33.860439   29      9  2021                   0   
2550189        33.860439   30      9  2021                   0   

         consecutive_HW_days  heat_wave_day  health_expend_p_capita  Asthma  \
0                          0              0             1164.962074    4.30   
1                          0              0             1164.962074    4.30   
2                          0              0             1164.962074    4.30   
3                          0              0             1164.962074    4.30   
4                          0              0             1164.962074    4.30   
...                      ...            ...                     ...     ...   
2550185                    0              0              814.707973    1.75   
2550186                    0              0              814.707973    1.75   
2550187                    0              0              814.707973    1.75   
2550188                    0              0              814.707973    1.75   
2550189                    0              0              814.707973    1.75   

         Chronic_depression  Chronic_lower_respiratory_diseases  \
0                      6.75                                 3.0   
1                      6.75                                 3.0   
2                      6.75                                 3.0   
3                      6.75                                 3.0   
4                      6.75                                 3.0   
...                     ...                                 ...   
2550185                1.25                                 1.6   
2550186                1.25       

In [24]:
# inspect shape
final_df.shape

(2550190, 62)

In [25]:
# drop the averages
final_df = final_df.drop(['wind_day_avg', 'dew_point_day_avg', 'uvb_4PM_avg', 'MRT_2AM_avg', 'MRT_4PM_avg', 'max_apparent_temperature_avg',
                          'min_apparent_temperature_avg', 'death_p_100k_avg'], 1)


In [26]:
# filter only instances that are part of a period with more than 3 consecutive days exceeding the 99th percentile
heatwave = final_df[final_df['consecutive_HW_days'] > 3]
# rename the previously obtained column comsum to heatwave_id
heatwave = heatwave.rename(columns={'cumsum': 'heatwave_id'})
heatwave['heatwave_id'] = heatwave['heatwave_id'].astype(str)
heatwave.head()

lat_long        date  longitude  latitude    uvb_4PM country  \
991   ('43.0', '-9.25')  2006-07-14      -9.25      43.0  295581.34      ES   
992   ('43.0', '-9.25')  2006-07-15      -9.25      43.0  288340.84      ES   
993   ('43.0', '-9.25')  2006-07-16      -9.25      43.0  302635.12      ES   
994   ('43.0', '-9.25')  2006-07-17      -9.25      43.0  249648.77      ES   
2050  ('43.0', '-9.25')  2013-07-05      -9.25      43.0  304197.66      ES   

     NUTS1 NUTS2  NUTS3  MRT_2AM  MRT_4PM Week_Number  99th_percentile  day  \
991    ES1  ES11  ES111    12.59    29.40         W28        25.658925   14   
992    ES1  ES11  ES111    14.20    28.87         W28        25.658925   15   
993    ES1  ES11  ES111    14.23    27.08         W28        25.658925   16   
994    ES1  ES11  ES111    13.51    32.36         W29        25.658925   17   
2050   ES1  ES11  ES111     9.85    24.10         W27        25.658925    5   

      month  year  threshold_exceeded  consecutive_HW_days  heat_wave_day  \
991       7  2006                   1                    4              1   
992       7  2006                   1                    4              1   
993       7  2006                   1                    4              1   
994       7  2006                   1                    4              1   
2050      7  2013                   1                    5              1   

      health_expend_p_capita  Asthma  Chronic_depression  \
991                  1769.92     4.3                6.75   
992                  1769.92     4.3                6.75   
993                  1769.92     4.3                6.75   
994                  1769.92     4.3                6.75   
2050                 1984.38     4.3                6.75   

      Chronic_lower_respiratory_diseases  \
991                                  3.0   
992                                  3.0   
993                                  3.0   
994                                  3.0   
2050                                 3.0   

      Coronary_heart_disease_or_angina_pectoris  Diabetes  \
991                                        0.75      7.15   
992                                        0.75      7.15   
993                                        0.75      7.15   
994                                        0.75      7.15   
2050                                       0.75      7.15   

      Heart_attack_or_chronic_consequences_of_heart_attack  \
991                                                0.85      
992                                                0.85      
993                                                0.85      
994                                                0.85      
2050                                               0.85      

      High_blood_pressure  Kidney_problems  Obesity  \
991                  19.0              2.6     16.1   
992                  19.0              2.6     16.1   
993                  19.0              2.6     16.1   
994                  19.0              2.6     16.1   
2050                 19.0              2.6     16.1   

      Stroke_or_chronic_consequences_of_stroke  week_nr NUTS2_Label  deaths  \
991                                        0.9  2006W28     Galicia   558.0   
992                                        0.9  2006W28     Galicia   558.0   
993                                        0.9  2006W28     Galicia   558.0   
994                                        0.9  2006W29     Galicia   607.0   
2050                                       0.9  2013W27     Galicia   665.0   

      population  density  population_65+  share_over_65  death_p_100k  \
991    2730107.0     92.5        585324.0      21.439599     20.438759   
992    2730107.0     92.5        585324.0      21.439599     20.438759   
993    2730107.0     92.5        585324.0      21.439599     20.438759   
994    2730107.0     92.5        585324.0      21.439599     22.233561   
2050   2761989.0     93.9        638250.0      23.108347    

In [27]:
# aggregate data by heatwave_id to get mean values of the numerical variables over that heat wave
heatwave = heatwave.groupby('heatwave_id').agg({'heatwave_id': 'first',
                                         'country': 'first',
                                         'NUTS1': 'first',
                                         'NUTS2': 'first',
                                         'NUTS2_Label': 'first',
                                         'lat_long': 'first',
                                         'date': ['min', 'max'],
                                         'latitude': 'mean',
                                         'longitude': 'mean',
                                         'min_apparent_temperature': 'mean',
                                         'min_apparent_temperature_vs_avg': 'mean',
                                         'max_apparent_temperature': 'mean',
                                         'max_apparent_temperature_vs_avg': 'mean',
                                         'uvb_4PM': 'mean',
                                         'uvb_4PM_vs_avg': 'mean',
                                         'wind_day': 'mean',
                                         'wind_day_vs_avg': 'mean',
                                         'dew_point_day': 'mean',
                                         'dew_point_day_vs_avg': 'mean',
                                         'MRT_2AM': 'mean',
                                         'MRT_2AM_vs_avg': 'mean',
                                         'MRT_4PM': 'mean',
                                         'MRT_4PM_vs_avg': 'mean',
                                         'NUTS2': 'first',
                                         'NUTS3': 'first',
                                         'Week_Number': 'first',
                                         '99th_percentile': 'mean',
                                         'month': 'mean',
                                         'year': 'first',
                                         'threshold_exceeded': 'sum',
                                         'consecutive_HW_days': 'max',
                                         'heat_wave_day': 'sum',
                                         'health_expend_p_capita': 'mean',
                                         'Asthma': 'mean',
                                         'Chronic_depression': 'mean',
                                         'Chronic_lower_respiratory_diseases': 'mean',
                                         'Coronary_heart_disease_or_angina_pectoris': 'mean',
                                         'Diabetes': 'mean',
                                         'Heart_attack_or_chronic_consequences_of_heart_attack': 'mean',
                                         'High_blood_pressure': 'mean',
                                         'Kidney_problems': 'mean',
                                         'Stroke_or_chronic_consequences_of_stroke': 'mean',
                                         'Obesity': 'mean',
                                         'week_nr': 'first',
                                         'deaths': 'mean',
                                         'population': 'mean',
                                         'share_over_65': 'mean',
                                         'density': 'mean',
                                         'death_p_100k': 'mean',
                                         'death_p_100k_vs_avg': 'mean',})
# inspect data
heatwave.head(7)

heatwave_id country NUTS1 NUTS2        NUTS2_Label  \
                  first   first first first              first   
heatwave_id                                                      
100                 100      PT   PT1  PT16        Centro (PT)   
10008             10008      CH   CH0  CH03    Nordwestschweiz   
10030             10030      CH   CH0  CH03    Nordwestschweiz   
1004               1004      PT   PT1  PT11              Norte   
10058             10058      CH   CH0  CH03    Nordwestschweiz   
10064             10064      CH   CH0  CH03    Nordwestschweiz   
10092             10092      CH   CH0  CH02  Espace Mittelland   

                       lat_long        date             latitude longitude  \
                          first         min         max     mean      mean   
heatwave_id                                                                  
100          ('39.25', '-9.25')  2018-08-02  2018-08-05    39.25     -9.25   
10008           ('47.5', '8.0')  2011-08-20  2011-08-23    47.50      8.00   
10030           ('47.5', '8.0')  2015-07-01  2015-07-07    47.50      8.00   
1004           ('41.5', '-7.0')  2018-08-02  2018-08-06    41.50     -7.00   
10058           ('47.5', '8.0')  2018-08-02  2018-08-06    47.50      8.00   
10064           ('47.5', '8.0')  2019-07-23  2019-07-26    47.50      8.00   
10092          ('46.75', '8.0')  2003-08-03  2003-08-13    46.75      8.00   

            min_apparent_temperature min_apparent_temperature_vs_avg  \
                                mean                            mean   
heatwave_id                                                            
100                        23.657500                        1.397669   
10008                      19.965000                        1.789727   
10030                      21.122857                        1.893521   
1004                       20.998000                        1.561944   
10058                      21.196000                        1.900078   
10064                      18.397500                        1.649211   
10092                      10.272727                        2.284291   

            max_apparent_temperature max_apparent_temperature_vs_avg  \
                                mean                            mean   
heatwave_id                                                            
100                        32.340000                        1.454426   
10008                      31.742500                        1.589615   
10030                      33.584286                        1.681849   
1004                       35.660000                        1.448005   
10058                      32.000000                        1.602510   
10064                      32.800000                        1.642573   
10092                      23.063636                        1.788616   

                   uvb_4PM uvb_4PM_vs_avg  wind_day wind_day_vs_avg  \
                      mean           mean      mean            mean   
heatwave_id                                                           
100          286961.135000       1.117221  0.976250        0.564086   
10008        147769.050000       1.184928  0.926875        0.616491   
10030        186158.305714       1.492763  1.072500        0.713350   
1004         281444.720000       1.206473  1.604000        0.903057   
10058        163035.102000       1.307343  1.076500        0.716011   
10064        187458.692500       1.503191  1.184375        0.787761   
10092        164366.635455       1.306543  0.512273        0.851105   

            dew_point_day dew_point_day_vs_avg    MRT_2AM MRT_2AM_vs_avg  \
                     mean                 mean       mean           mean   
heatwave_id                                                                
100             16.538750             1.148659  15.037500       1.497314   
10008           16.796875             1.488345  12.290000       1.876485   
10030           17.035714            

In [28]:
# reduce column index levels to 1
heatwave.columns = heatwave.columns.get_level_values(0)
heatwave.head(1)

heatwave_id country NUTS1 NUTS2  NUTS2_Label            lat_long  \
heatwave_id                                                                    
100                 100      PT   PT1  PT16  Centro (PT)  ('39.25', '-9.25')   

                   date        date  latitude  longitude  \
heatwave_id                                                
100          2018-08-02  2018-08-05     39.25      -9.25   

             min_apparent_temperature  min_apparent_temperature_vs_avg  \
heatwave_id                                                              
100                           23.6575                         1.397669   

             max_apparent_temperature  max_apparent_temperature_vs_avg  \
heatwave_id                                                              
100                             32.34                         1.454426   

                uvb_4PM  uvb_4PM_vs_avg  wind_day  wind_day_vs_avg  \
heatwave_id                                                          
100          286961.135        1.117221   0.97625         0.564086   

             dew_point_day  dew_point_day_vs_avg  MRT_2AM  MRT_2AM_vs_avg  \
heatwave_id                                                                 
100               16.53875              1.148659  15.0375        1.497314   

             MRT_4PM  MRT_4PM_vs_avg  NUTS3 Week_Number  99th_percentile  \
heatwave_id                                                                
100          36.3975        1.235334  PT16B         W31        30.174927   

             month  year  threshold_exceeded  consecutive_HW_days  \
heatwave_id                                                         
100            8.0  2018                   4                    4   

             heat_wave_day  health_expend_p_capita  Asthma  \
heatwave_id                                                  
100                      4                 1878.02    5.55   

             Chronic_depression  Chronic_lower_respiratory_diseases  \
heatwave_id                                                           
100                       12.05                                5.65   

             Coronary_heart_disease_or_angina_pectoris  Diabetes  \
heatwave_id                                                        
100                                               4.15      9.65   

             Heart_attack_or_chronic_consequences_of_heart_attack  \
heatwave_id                                                         
100                                                       1.55      

             High_blood_pressure  Kidney_problems  \
heatwave_id                                         
100                        25.95             4.45   

             Stroke_or_chronic_consequences_of_stroke  Obesity  week_nr  \
heatwave_id                                                               
100                                               1.9     16.9  2018W31   

             deaths  population  share_over_65  density  death_p_100k  \
heatwave_id                                                             
100           540.0   2231346.0      23.987539     79.5     24.200639   

             death_p_100k_vs_avg  
heatwave_id                       
100                     1.124589

In [29]:
# inspect shape
heatwave.shape

(3350, 50)

In [29]:
# write data to csv
heatwave.to_csv('heatwave1.csv')

## Used in Notebook of the modelling part --> Modelling.ipnyb

# END